// nylas/lib/cjs/models/webhooks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookTriggers = void 0;
/**
 * Enum representing the available webhook triggers.
 */
var WebhookTriggers;
(function (WebhookTriggers) {
    WebhookTriggers["CalendarCreated"] = "calendar.created";
    WebhookTriggers["CalendarUpdated"] = "calendar.updated";
    WebhookTriggers["CalendarDeleted"] = "calendar.deleted";
    WebhookTriggers["EventCreated"] = "event.created";
    WebhookTriggers["EventUpdated"] = "event.updated";
    WebhookTriggers["EventDeleted"] = "event.deleted";
    WebhookTriggers["GrantCreated"] = "grant.created";
    WebhookTriggers["GrantUpdated"] = "grant.updated";
    WebhookTriggers["GrantDeleted"] = "grant.deleted";
    WebhookTriggers["GrantExpired"] = "grant.expired";
    WebhookTriggers["MessageSendSuccess"] = "message.send_success";
    WebhookTriggers["MessageSendFailed"] = "message.send_failed";
})(WebhookTriggers = exports.WebhookTriggers || (exports.WebhookTriggers = {}));
