// nylas/lib/esm/resources/redirectUris.js
import { Resource } from './resource.js';
/**
 * A collection of redirect URI related API endpoints.
 *
 * These endpoints allows for the management of redirect URIs.
 */
export class RedirectUris extends Resource {
    /**
     * Return all Redirect URIs
     * @return The list of Redirect URIs
     */
    list({ overrides } = {}) {
        return super._list({
            overrides,
            path: '/v3/applications/redirect-uris',
        });
    }
    /**
     * Return a Redirect URI
     * @return The Redirect URI
     */
    find({ redirectUriId, overrides, }) {
        return super._find({
            overrides,
            path: `/v3/applications/redirect-uris/${redirectUriId}`,
        });
    }
    /**
     * Create a Redirect URI
     * @return The created Redirect URI
     */
    create({ requestBody, overrides, }) {
        return super._create({
            overrides,
            path: '/v3/applications/redirect-uris',
            requestBody,
        });
    }
    /**
     * Update a Redirect URI
     * @return The updated Redirect URI
     */
    update({ redirectUriId, requestBody, overrides, }) {
        return super._update({
            overrides,
            path: `/v3/applications/redirect-uris/${redirectUriId}`,
            requestBody,
        });
    }
    /**
     * Delete a Redirect URI
     * @return The deleted Redirect URI
     */
    destroy({ redirectUriId, overrides, }) {
        return super._destroy({
            overrides,
            path: `/v3/applications/redirect-uris/${redirectUriId}`,
        });
    }
}
