// nylas/lib/cjs/resources/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Events = void 0;
const resource_js_1 = require("./resource.js");
/**
 * Nylas Events API
 *
 * The Nylas Events API allows you to create, update, and delete events on user calendars.
 */
class Events extends resource_js_1.Resource {
    /**
     * Return all Events
     * @return The list of Events
     */
    list({ identifier, queryParams, overrides, }) {
        return super._list({
            queryParams,
            path: `/v3/grants/${identifier}/events`,
            overrides,
        });
    }
    /**
     * Return an Event
     * @return The Event
     */
    find({ identifier, eventId, queryParams, overrides, }) {
        return super._find({
            path: `/v3/grants/${identifier}/events/${eventId}`,
            queryParams,
            overrides,
        });
    }
    /**
     * Create an Event
     * @return The created Event
     */
    create({ identifier, requestBody, queryParams, overrides, }) {
        return super._create({
            path: `/v3/grants/${identifier}/events`,
            queryParams,
            requestBody,
            overrides,
        });
    }
    /**
     * Update an Event
     * @return The updated Event
     */
    update({ identifier, eventId, requestBody, queryParams, overrides, }) {
        return super._update({
            path: `/v3/grants/${identifier}/events/${eventId}`,
            queryParams,
            requestBody,
            overrides,
        });
    }
    /**
     * Delete an Event
     * @return The deletion response
     */
    destroy({ identifier, eventId, queryParams, overrides, }) {
        return super._destroy({
            path: `/v3/grants/${identifier}/events/${eventId}`,
            queryParams,
            overrides,
        });
    }
    /**
     * Send RSVP. Allows users to respond to events they have been added to as an attendee.
     * You cannot send RSVP as an event owner/organizer.
     * You cannot directly update events as an invitee, since you are not the owner/organizer.
     * @return The send-rsvp response
     */
    sendRsvp({ identifier, eventId, requestBody, queryParams, overrides, }) {
        return super._create({
            path: `/v3/grants/${identifier}/events/${eventId}/send-rsvp`,
            queryParams,
            requestBody,
            overrides,
        });
    }
}
exports.Events = Events;
