// nylas/lib/cjs/resources/threads.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Threads = void 0;
const resource_js_1 = require("./resource.js");
/**
 * Nylas Threads API
 *
 * The Nylas Threads API allows you to list, find, update, and delete threads on user accounts.
 */
class Threads extends resource_js_1.Resource {
    /**
     * Return all Threads
     * @return A list of threads
     */
    list({ identifier, queryParams, overrides, }) {
        return super._list({
            queryParams,
            overrides,
            path: `/v3/grants/${identifier}/threads`,
        });
    }
    /**
     * Return a Thread
     * @return The thread
     */
    find({ identifier, threadId, overrides, }) {
        return super._find({
            path: `/v3/grants/${identifier}/threads/${threadId}`,
            overrides,
        });
    }
    /**
     * Update a Thread
     * @return The updated thread
     */
    update({ identifier, threadId, requestBody, overrides, }) {
        return super._update({
            path: `/v3/grants/${identifier}/threads/${threadId}`,
            requestBody,
            overrides,
        });
    }
    /**
     * Delete a Thread
     * @return The deleted thread
     */
    destroy({ identifier, threadId, overrides, }) {
        return super._destroy({
            path: `/v3/grants/${identifier}/threads/${threadId}`,
            overrides,
        });
    }
}
exports.Threads = Threads;
